FEATURE: New workspace module (fusion & htmx) v1

Sebobo opened the pull request:
Resolves partly: https://github.com/neos/neos-development-collection/issues/4255

A new HTMX based UI with improved usability for the workspace overview based on the package https://github.com/Sebobo/Shel.Neos.WorkspaceModule.

This change is the first milestone and rewrites the via https://github.com/neos/neos-development-collection/pull/5118 into Neos.Workspace.Ui extracted fluid templates to Fusion and introduces the usage of HTMX. 

**Upgrade instructions:**

If any part of the module was overwritten in a project, the change will break any of those adjustments.
But the module is now easier to adjust from the outside than before as it uses Fusion based rendering which can be extended by including additional Fusion paths via a `Views.yaml`.


**Workspace disappears after changing visibility**

This can happen for previously created workspaces or if the workspace was manipulated via CLI too.
For that case look into

```sh
./flow workspace:show
```

If it says "There are no role assignments" or if no "USER" has the "ROLE" set to "MANAGER", then assign a user this workspace to make it effectively working (and private) again:

```sh
./flow workspace:assignrole <workspace> <user name> manager --type user
```

Sebobo committed:
FEATURE: New workspace overview UX

Sebobo committed:
TASK: Improve workspace column styling

Sebobo committed:
TASK: Implement workspace delete modal as popover

Sebobo committed:
TASK: Extract review script from show template

Sebobo committed:
TASK: Refactor create workspace dialog to Fusion

Sebobo committed:
TASK: Improve new workspace template

Sebobo committed:
FEATURE: Refactor edit template to AFX

Sebobo committed:
FEATURE: Introduce htmx for workspace module delete action

Sebobo committed:
TASK: Extract htmxconfig into component

Sebobo committed:
TASK: Improve HTMX integration in workspace UI

Use http headers for proper status codes and
transport flash messages

Sebobo committed:
BUGFIX: Use new workspace ui privilege names in userservice

Sebobo committed:
TASK: Sort base workspace options

Sebobo committed:
TASK: Improve workspace creation notifications

Sebobo committed:
FEATURE: Turn workspace create dialog into htmx popover

Sebobo committed:
FEATURE: Render number of workspaces by visibility in module footer

Sebobo committed:
TASK: Generate delete modal for workspaces on demand

Sebobo committed:
TASK: Move purely presentational components

Sebobo committed:
TASK: Add button component for workspace ui

Sebobo committed:
TASK: Refactor workspace edit action to a modal

Sebobo committed:
TASK: Refactor workspace count to component

Sebobo committed:
TASK: Render workspace list after workspace has been updated without redirect

Sebobo committed:
FEATURE: Render changes countsin workspace list

Sebobo committed:
BUGFIX: Make editing of workspaces possible again

Sebobo committed:
FEATURE: Localise workspace state

Sebobo committed:
BUGFIX: Solve view error with workspace review

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/4255-new-workspace-management-ui

mhsdesign committed:
WIP: Fix php integration to use new workspace permissions and metadata

phpstan is happy and fusion boots

mhsdesign committed:
WIP: Rename use `WorkspaceListItem` instead of `WorkspaceDetails`

mhsdesign committed:
WIP: Pass `userWorkspaceName` instead of full workspace

mhsdesign committed:
WIP: Remove occurrence of `.workspace` in show action

We dont pass the full workspace object anymore

mhsdesign committed:
WIP: Remove old view models

pKallert committed:
Feature: Move a little bit of review to fusion

pKallert committed:
Fix: Naming of workspaces in overview

pKallert committed:
Merge branch 'feature/4255-new-workspace-management-ui' into feature/new-workspace-review

pKallert committed:
Feature: Apply stash

JamesAlias committed:
TASK: Workspace management edit action

pKallert committed:
Feature: finish change diff

pKallert committed:
Merge in remote

pKallert committed:
Feature: merge in current branch

pKallert committed:
Feature: Add styling

pKallert committed:
Feature: Add reload after discard

pKallert committed:
Feature: discard action

JamesAlias committed:
TASK: Workspace management create and delete action

JamesAlias committed:
TASK: Fix EditWorkspaceDto

pKallert committed:
Merge branch 'feature/4255-new-workspace-management-ui' of https://github.com/neos/neos-development-collection into feature/4255-new-workspace-management-ui

pKallert committed:
Merge branch 'feature/4255-new-workspace-management-ui' into feature/new-workspace-review

JamesAlias committed:
TASK: change baseWorkspace in edit workspace

pKallert committed:
Feature: Add action footer back

pKallert committed:
Feature: remove unused class

pKallert committed:
Merge pull request #5311 from pKallert/feature/new-workspace-review

FEATURE: convert detail view to fusion

pKallert committed:
Feature: Make publish Documents work

pKallert committed:
Feature: Remove single node publish

JamesAlias committed:
TASK: remove owner column in workspace management table, fix some permissions

JamesAlias committed:
TASK: fix icons & ignore live workspace in workspace listitems

pKallert committed:
Feature: Add Javascript functionality

pKallert committed:
Merge pull request #5319 from pKallert/feature/add-review-functionality

FEATURE: Add Buttons and Javascript functions for review

pKallert committed:
Feature: Adjust design of review page

pKallert committed:
Feature: Adjust permissions

pKallert committed:
Merge pull request #5323 from pKallert/feature/redesign-review-page

Feature/redesign review page

pKallert committed:
Feature: Add DTO for viewmodel

JamesAlias committed:
WIP: WorkspaceRoleAssignment edit

pKallert committed:
Feature: Adjust SiteChanges

pKallert committed:
Feature: Add tag logic

pKallert committed:
Merge pull request #5326 from pKallert/feature/review-page-dto

Feature: Add dto for review

JamesAlias committed:
WIP: WorkspaceRoleAssignment edit, add, delete

mhsdesign commented:
Annnd the new cr selector got lost during the rewrite:

<img width="657" alt="image" src="https://github.com/user-attachments/assets/65b013c2-5a14-41d7-bd47-a9372f2c05ea">

JamesAlias committed:
TASK: Fix delete workspace, render workspace count oob

mhsdesign commented:
FYI i moved all adjustments don to the `WorkspaceService` into this separate PR https://github.com/neos/neos-development-collection/pull/5334 so we can discuss them and also write tests for them etc ...

When that is merged ill update this pr

JamesAlias committed:
TASK: Fix edit workspace pattern, but user help still needed

JamesAlias committed:
TASK: Fix edit & remove WorkspaceRoleAssignment

pKallert committed:
Feature: Merge in latest 9.0 branch

pKallert committed:
Feature: Add permission queries to review page

Sebobo committed:
TASK: Flip cancel and confirm buttons and adjust position in modals

Sebobo committed:
TASK: Add basic user stories for the workspace ui

Sebobo committed:
BUGFIX: Show back button in workspace review when there are no changes

pKallert committed:
Feature: Consider parents and children when using checkboxes

pKallert committed:
Merge branch 'feature/4255-new-workspace-management-ui' of https://github.com/neos/neos-development-collection into feature/4255-new-workspace-management-ui

pKallert committed:
Merge branch '9.0' into feature/4255-new-workspace-management-ui

pKallert committed:
Task: Move removeWorkspaceRoleAssignment and deleteWorkspaceMetadata to metadataAndRoleRepository

JamesAlias committed:
WIP: WorkspaceRoleAssignment edit, add, delete

pKallert committed:
Feature: Add publishAll and DiscardAll functions

pKallert committed:
Merge branch 'feature/4255-new-workspace-management-ui' of https://github.com/neos/neos-development-collection into feature/4255-new-workspace-management-ui

pKallert committed:
Feature: Make Discard and publish selected changes work

pKallert committed:
Feature: Separate changes in doucment into different dimensions

pKallert committed:
Feature: Change folder structure

pKallert committed:
Fix: Make menu work on review page

pKallert committed:
Fix: Show the right labels for properties

Sebobo committed:
Merge branch '9.0' into feature/4255-new-workspace-management-ui

Sebobo committed:
TASK: Change fusion folder structure

Sebobo committed:
BUGFIX: Resolve nested swap issues

Sebobo committed:
module wrap

Sebobo committed:
TASK: Introduce module wrapper for workspace module views

Sebobo committed:
BUGFIX: Show flash message on workspace delete

Sebobo committed:
TASK: Reformat workspace module js

bwaidelich commented on Neos.Workspace.Ui/Classes/ViewModel/ChangeItem.php:
```diff
@@ -0,0 +1,36 @@
+        public ?string                 $createdDateTime,
+        public string                  $label,
+        public string                  $icon,
+        public ContentChangeItems      $contentChanges,
```
Nitpick, but we don't really do these aligned arguments any more, because they don't really increase readability and lead to larger diffs when a new param is introduced that exceeds the previous max length
```suggestion
        public string $serializedNodeAddress,
        public bool $hidden,
        public bool $isRemoved,
        public bool $isNew,
        public bool $isMoved,
        public array $dimensions,
        public ?string $lastModificationDateTime,
        public ?string $createdDateTime,
        public string $label,
        public string $icon,
        public ContentChangeItems $contentChanges,
```

bwaidelich commented on Neos.Workspace.Ui/Classes/ViewModel/ConfirmDeleteWorkspaceRoleAssignmentFormData.php:
```diff
@@ -0,0 +1,32 @@
+#[Flow\Proxy(false)]
+final readonly class ConfirmDeleteWorkspaceRoleAssignmentFormData
+{
+    public function __construct(
+        public WorkspaceName  $workspaceName,
+        public WorkspaceTitle $workspaceTitle,
+        public string         $subjectValue,
+        public string         $subjectType,
+    )
+    {
+    }
```
```suggestion
    public function __construct(
        public WorkspaceName $workspaceName,
        public WorkspaceTitle $workspaceTitle,
        public string $subjectValue,
        public string $subjectType,
    ) {
    }
```

bwaidelich commented on Neos.Workspace.Ui/Classes/ViewModel/CreateWorkspaceRoleAssignmentFormData.php:
```diff
@@ -0,0 +1,59 @@
+final readonly class CreateWorkspaceRoleAssignmentFormData
+{
+    public function __construct(
+        public WorkspaceName  $workspaceName,
+        public WorkspaceTitle $workspaceTitle,
+        /**
+         * Options for the workspaceManager selector where the key is the user identifier and the value is the user name.
+         * @var array<string, string>
+         */
+        public array          $userOptions,
+        /**
+         * Options for the workspaceManager selector where the value is the group.
+         * @var array<string>
+         */
+        public array          $groupOptions,
+        /**
+         * TODO: translate subject type labels?
+         * Options for the workspaceManager selector where the key is the subject type and the value is the subject label.
+         * @var array<string>
+         */
+        public array          $subjectTypeOptions,
+        /**
+         * TODO: translate role labels?
+         * Options for the workspaceManager selector where the key is the role identifier and the value is the role label.
+         * @var array<string, string>
+         */
+        public array          $roleOptions
```
```suggestion
        public WorkspaceName $workspaceName,
        public WorkspaceTitle $workspaceTitle,
        /**
         * Options for the workspaceManager selector where the key is the user identifier and the value is the user name.
         * @var array<string, string>
         */
        public array $userOptions,
        /**
         * Options for the workspaceManager selector where the value is the group.
         * @var array<string>
         */
        public array $groupOptions,
        /**
         * TODO: translate subject type labels?
         * Options for the workspaceManager selector where the key is the subject type and the value is the subject label.
         * @var array<string>
         */
        public array $subjectTypeOptions,
        /**
         * TODO: translate role labels?
         * Options for the workspaceManager selector where the key is the role identifier and the value is the role label.
         * @var array<string, string>
         */
        public array $roleOptions
```

bwaidelich commented on Neos.Workspace.Ui/Classes/ViewModel/EditWorkspaceFormData.php:
```diff
@@ -0,0 +1,39 @@
+#[Flow\Proxy(false)]
+final readonly class EditWorkspaceFormData
+{
+    public function __construct(
+        public WorkspaceName        $workspaceName,
+        public WorkspaceTitle       $workspaceTitle,
+        public WorkspaceDescription $workspaceDescription,
+        public bool                 $workspaceHasChanges,
+        public WorkspaceName        $baseWorkspaceName,
+        /**
+         * Options for the baseWorkspace selector where the key is the workspace name and the value is the workspace title.
+         * @var array<string, string>
+         */
+        public array                $baseWorkspaceOptions,
+    )
+    {
```
doc comments for promoted properties must be put `@param` not `@var` annotations

```suggestion
    /**
     * @param array<string, string> $baseWorkspaceOptions Options for the baseWorkspace selector where the key is the workspace name and the value is the workspace title.
     */
    public function __construct(
        public WorkspaceName $workspaceName,
        public WorkspaceTitle $workspaceTitle,
        public WorkspaceDescription $workspaceDescription,
        public bool $workspaceHasChanges,
        public WorkspaceName $baseWorkspaceName,
        public array $baseWorkspaceOptions,
    ) {
```

bwaidelich commented on Neos.Workspace.Ui/Classes/ViewModel/EditWorkspaceRoleAssignmentsFormData.php:
```diff
@@ -0,0 +1,42 @@
+#[Flow\Proxy(false)]
+final readonly class EditWorkspaceRoleAssignmentsFormData
+{
+    public function __construct(
+        public WorkspaceName  $workspaceName,
+        public WorkspaceTitle $workspaceTitle,
+        public bool           $roleAssignmentsEditable,
+        /**
+         * @var array<RoleAssignmentListItem>
+         */
+        public array          $roleAssignments,
+    )
+    {
```
```suggestion
    /**
     * @param array<RoleAssignmentListItem> $roleAssignments
     */
    public function __construct(
        public WorkspaceName $workspaceName,
        public WorkspaceTitle $workspaceTitle,
        public bool $roleAssignmentsEditable,
        public array $roleAssignments,
    ) {
```

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -81,6 +83,12 @@ public function setWorkspaceTitle(ContentRepositoryId $contentRepositoryId, Work
         $this->requireManagementWorkspacePermission($contentRepositoryId, $workspaceName);
         $workspace = $this->requireWorkspace($contentRepositoryId, $workspaceName);
+        $workspaceMetadata = $this->metadataAndRoleRepository->loadWorkspaceMetadata($contentRepositoryId, $workspaceName);
+
+        // WHY: Do not update if the title is the same
+        if ($workspaceMetadata->title->equals($newWorkspaceTitle)) {
+            return;
+        }
```
My bad, but this check is obsolete with #5397

```suggestion
```

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -91,6 +99,11 @@ public function setWorkspaceDescription(ContentRepositoryId $contentRepositoryId
         $this->requireManagementWorkspacePermission($contentRepositoryId, $workspaceName);
         $workspace = $this->requireWorkspace($contentRepositoryId, $workspaceName);
+        $workspaceMetadata = $this->metadataAndRoleRepository->loadWorkspaceMetadata($contentRepositoryId, $workspaceName);
+        // WHY: Do not update if the title is the same
+        if ($workspaceMetadata->description->equals($newWorkspaceDescription)) {
+            return;
+        }
```
obsolete with #5397

```suggestion
```

bwaidelich reviewed (commented):
Just some, partly nitpicky, comments :)

mhsdesign commented:
I corrected the workspace service so it works more like anticipated (and like in 8.3) in https://github.com/neos/neos-development-collection/pull/5334

One thing I dont understand is if we really want the change that a neos administrator can read, write and mange every workspace? 

```php
        if ($userIsAdministrator) {
            return WorkspacePermissions::all('User is Neos Administrator');
        }
```

That was change by @JamesAlias see https://github.com/neos/neos-development-collection/commit/916e51009e9c97b42aa32372b94d149a2022faeb

Currently admins only are managers (and cannot read the workspace) unless they promote themselves to anything else - though i see that might be the logic flaw.

EDIT: We reverted that change as admins are only allowed to be managers as escape hatch

pKallert committed:
Fix: Do not show duplicate footer on discard/publish single Document

mhsdesign committed:
Adjust to PSR12 code style for parameter spacing

Co-authored-by: Bastian Waidelich <bastian@neos.io>

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/4255-new-workspace-management-ui

mhsdesign committed:
TASK: Adjustments after workspace service fixes

see https://github.com/neos/neos-development-collection/pull/5334

mhsdesign commented:
Okay, i extracted all the workspace service adjustments and wrote tests that it works now: https://github.com/neos/neos-development-collection/pull/5334 indeed some parts were odd and needed fixing.

That is merged into 9.0 and i also merged that back into there and adjusted this pr to the new method signature.

That means the ci should have a chance to get this green soonish as there are no core changes anymore.
Only that it seems all the new code is not really written with phpstan :D It really helps to run

```
composer run lint:phpstan
```

every once in a while ^^

Currently there a lot of possible null access bugs.

mhsdesign committed:
TASK: Add `equals` helper to subtree tags

mhsdesign committed:
BUGFIX: Only compare actual node tags (without inheritance)

and use introduced native diff util.
Further fixes as `equals` will also be true now if the order is different.

mhsdesign committed:
TASK: Basic phpstan adjustments for array types and nullability

mhsdesign committed:
TASK: Adjust code to pass user to `getWorkspaceListItems` instead

mhsdesign committed:
TASK: Further phpstan adjustments

(and use hash to index dimensions)

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into feature/4255-new-workspace-management-ui

mhsdesign committed:
TASK: Revert WIP workspace role assigment Ui

(sorry robert :D - well work on this in a another pr - by reverting this change ^^)

reverts all actions:

- editWorkspaceRoleAssignmentsAction
- createWorkspaceRoleAssignmentAction
- addWorkspaceRoleAssignmentAction
- confirmDeleteWorkspaceRoleAssignmentAction
- deleteWorkspaceRoleAssignmentAction

mhsdesign commented:
As discussed in the daily today i proposed to revert the role assignment ui part instead for this pr and have it done separately:

https://github.com/neos/neos-development-collection/pull/5132/commits/4aed44146bf8de69d6b0809c77b6e110d2409712

i argued against a feature flag, because some buttons are currently not working for the assignments and i hope that when reintroducing this change in a new pr the code can be better reviewed and tested.

Also by fully removing this code for now we also keep this pr slimmer and more easy to review what is great as well.

reverts all actions:

- editWorkspaceRoleAssignmentsAction
- createWorkspaceRoleAssignmentAction
- addWorkspaceRoleAssignmentAction
- confirmDeleteWorkspaceRoleAssignmentAction
- deleteWorkspaceRoleAssignmentAction

(sorry robert :D)

-----

one more thing: Sebastian will attempt to reimplement a basic kind of access control again (basically mapping the 'old' internal / private functionality on the new model) - but i would be fine to do this in a followup pr too.

Currently the Neos beta contains the Neos 8 workspace module in super slimmed down and broken, having this pr merged - even without any acl (which is NOT implemented in the current beta) would already be a step forward. And merging this right now would give us maybe still the chance to get the reverted full acl (edit role assignments) running.
In case we dont get the full acl working soon, we should then at least ship with Neos 9.0 the old internal & private functionality.

-----

So in order to get this merged id say we can begin with the bug hunting and all features are followups?
I have a bug to begin: The cancel button (for creation) doesnt work.

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -422,14 +499,9 @@ public function rebaseAndRedirectAction(string $targetNode, Workspace $targetWor
         /** @todo do something else
```
whatever this did before (possibly in 8.3) we should reimplement that

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -612,41 +743,37 @@ public function publishWorkspaceAction(WorkspaceName $workspace): void
-                }
-            }
+        foreach($this->getChangesFromWorkspace($selectedWorkspace, $contentRepository) as $change) {
+            if($change->deleted) $changesCount['removed']++;
```
this codestyle is not psr12 we always use braces

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -656,27 +783,24 @@ protected function computePendingChanges(Workspace $selectedWorkspace, ContentRe
-            $subgraph = $contentGraph->getSubgraph(
-                $change->originDimensionSpacePoint?->toDimensionSpacePoint() ?: $arbitraryDimensionSpacePoint,
+            if ($change->originDimensionSpacePoint === null) {
+                // todo implement support for change node type!!! Because originDimensionSpacePoint is null currently for that case.
```
this might be solvable by https://github.com/neos/neos-development-collection/pull/5395 ... lets see

mhsdesign replied:
hotfixed via https://github.com/neos/neos-development-collection/pull/5132/commits/c4eb48fb00f72f5f88af74f7368c6a2d13e477d1

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -942,7 +1133,29 @@ protected function getPropertyLabel($propertyName, Node $changedNode)
             return $propertyName;
         }
-        return $properties[$propertyName]['ui']['label'];
+        $packageKey = 'Neos.Neos';
```
why is this translation logic needed here? its duplicated from `\Neos\ContentRepositoryRegistry\Configuration\NodeTypeEnrichmentService::splitIdentifier`

and was introduced via b830ee8a2d88be1d6cc81c41667c85f1ddb5dcdd

pKallert replied:
Could not find a better way to get the translations for the property labels here when using i18n - suggestions are welcome 😄

mhsdesign replied:
I tested it now with `TranslationHelper()->translate()` and it works still. Its also a little hacky but less code duplication and we do it in other places too.

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -982,39 +1195,57 @@ protected function postProcessDiffArray(array &$diffArray): void
             $baseWorkspaceOptions[$workspace->workspaceName->value] = $workspaceMetadata->title->value;
         }
 
+        // Sort the base workspaces by title, but make sure the live workspace is always on top
+        uksort($baseWorkspaceOptions, static function (string $a, string $b) {
+            if ($a === 'live') {
+                return -1;
+            }
+            if ($b === 'live') {
+                return 1;
+            }
+            return strcasecmp($a, $b);
+        });
```
this is kinda odd, but makes sense i guess?

Introduced via 384b6e2ebfcfc30e618cb16c0e085636b9dc487a

Sebobo replied:
With many workspaces it was uncomfortable to search for the live workspace in the dropdown, which is the most used one.

mhsdesign commented on Neos.Workspace.Ui/Classes/Mvc/HtmxRequestPattern.php:
```diff
@@ -0,0 +1,26 @@
+use Neos\Flow\Mvc\ActionRequest;
+use Neos\Flow\Security\RequestPatternInterface;
+
+final class HtmxRequestPattern implements RequestPatternInterface
```
for "framework" like things like these, maybe we should add a little docs why this is needed? This will possibly be moved to Flow or Neos at some point right?

Sebobo replied:
Yes as this can affect integrator code too.

mhsdesign replied:
we discussed that if we want to support htmx for other modules as well we will probably create a dedicated framework code package for that. But for now its easiest to have it all in the ui code package and restrict it to only affect this package.

Sebobo replied:
We did that yesterday already :)

mhsdesign commented on Neos.Workspace.Ui/Classes/ViewModel/ContentChanges/ContentChange.php:
```diff
@@ -0,0 +1,22 @@
+use Neos\Flow\Annotations as Flow;
+
+#[Flow\Proxy(false)]
+readonly class ContentChange
```
i guess this should be either abstract or an interface?, or not needed at all if we uses unions

mhsdesign replied:
Fixed via https://github.com/neos/neos-development-collection/pull/5132/commits/0e17338d5206b99b848b3d50e107cc3ebe23628a

mhsdesign commented on Neos.Workspace.Ui/Documentation/UserStories.rst:
this list needs to be carefully reviewed if its still up to date with the Neos 9 workspace implementation.

Especially there is no definition for shared or internal:

> The visibility of the workspace (e.g. "Shared", "Private") and its access control list (ACL)

and they are no concepts for the current model either

--------

in general this readme might be better off as functional test - not using the workspace ui - but by testing the flow actions?

Behaviour in delete workspace is already different, as workspace with changes cannot be deleted see `throwStatus(403, 'Workspace has unpublished nodes');`.

So before merge, this list needs to be up to date or enforced, or we put this into a discuss post or something were we rather discuss the behaviour:)

mhsdesign commented on Neos.Workspace.Ui/Resources/Private/Fusion/Features/Workspace/Modals/Edit.fusion:
```diff
@@ -0,0 +1,128 @@
+  baseWorkspaceName = ''
+
+  workspaceHasChanges = false
+  /// array<workspaceName: string, workspaceTitle: string>
```
sidetrack: spaces are not allowed here technically but we probably need to refine the proposal https://discuss.neos.io/t/documenting-fusion/6056/35

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -136,85 +158,88 @@ public function indexAction(): void
         $this->view->assign('displayContentRepositorySelector', $numberOfContentRepositories > 1);
```
this selector, and the ability to choose between content repositories, got lost (see https://github.com/neos/neos-development-collection/pull/5132/#issuecomment-2447319888)

we can probably live without this. Though we then should fully remove the unused variables?

mhsdesign reviewed (changes requested):
okay i had a dedicated look at the php and added  a few comments i like to discuss.

Additionally there are 25!!! NEW todos in the code, its obvious that some todos will be unresolved, but lets have a look at those - and maybe remove todos already that are none?

(quick way to find todos:)
```
git diff -U0 origin/9.0...HEAD > ../myCurrent.diff
```
search `myCurrent.diff` for `^\+.*todo` :D

Sebobo committed:
TASK: Translate cancel labels and autofocus primary modal actions

Sebobo committed:
BUGFIX: Convert review actions to htmx and make flash messages work

Sebobo committed:
BUGFIX: Hide emtpy diffs

Sebobo committed:
TASK: Code cleanup

Sebobo committed:
BUGFIX: Add space between icon in button and its label

Sebobo committed:
BUGFIX: Reload review correctly after discard or publish workspace

Sebobo committed:
FEATURE: Set basic visibility on workspace creation

mhsdesign commented on Neos.Workspace.Ui/Classes/Controller/WorkspaceController.php:
```diff
@@ -185,7 +185,8 @@ public function reviewAction(WorkspaceName $workspace): void
                 Message::SEVERITY_ERROR
             );
-            $this->redirect('index');
+            $this->indexAction();
```
regarding all changes in e94d23a22fab844d5921a446bc640a4d1f4bcce9 we can instead of calling the other action directly use flow `$this->forward('index')` which does basically that but in a "correct" way

Sebobo replied:
Done

mhsdesign commented on Neos.Workspace.Ui/Resources/Private/Fusion/Features/Review/Actions/Redirects.fusion:
```diff
@@ -0,0 +1,10 @@
+# Actions that update a workspace and redirect to the review view
+
+# Discard actions
+Neos.Workspace.Ui.WorkspaceController.discardWorkspace < Neos.Workspace.Ui.WorkspaceController.review
```
copy path in fusion is deprecated and its not a performant behaviour (as it literally copies the whole thing over and over during parsing for the runtime)

I removed all occurrences from the code base and dont think its necessary here.

How about
A: Use a dedicated prototype
B: instead of using the default delegating behaviour of the FusionView, set the entry path to `Neos.Workspace.Ui.WorkspaceController` and there use a match which uses `request.controllerActionName`

Sebobo replied:
We don't need this if `forward` works properly

Sebobo replied:
Yes I removed all of it again :)

Sebobo committed:
TASK: Cleanup action redirects by using forward

Sebobo committed:
TASK: Remove last modification date column until we can have the date available in the api

Sebobo committed:
BUGFIX: Reload workspace list after creation

Sebobo committed:
FEATURE: Sort workspaces by title

Sebobo committed:
BUGFIX: Show correct workspace visibility state in the workspace list

Sebobo committed:
TASK: Only match htmx pattern for workspace module

Sebobo committed:
BUGFIX: Disable modifying personal workspaces in workspace list

Sebobo committed:
FEATURE: Allow modifying basic workspace visibility in edit dialog

pKallert commented:
Created PR for updating URL for new review in Neos UI: https://github.com/neos/neos-ui/pull/3904

Sebobo committed:
TASK: Use button disabled prop

Sebobo committed:
TASK: Solve todos from WorkspaceListItem

Sebobo committed:
TASK: Remove broken rebaseAndRedirect functionality

We have to find out if we want to keep the feature and
implement it in a more sensible way

Sebobo committed:
BUGFIX: Hide empty tag changes and improve label

Sebobo committed:
TASK: Use workspace classification to handle personal workspaces in the ui

Sebobo committed:
TASK: Remove unfinished CR switch functionality

Sebobo committed:
TASK: Remove user stories

Sebobo committed:
TASK: Solve code style issue

mhsdesign commented:
I extracted the obsolete policy yaml removals into 

- https://github.com/neos/neos-development-collection/pull/5417
- https://github.com/neos/neos-development-collection/pull/5418

as i felt they were to big to include here and deserve some additional information.

They can be merged later.

mhsdesign committed:
BUGFIX: Handle NodeAggregateNameWasChanged or NodeAggregateTypeWasChanged changes

This hack originally made in 4fe6b983a32bf559c6a78f5e74038c4d9de16a78 got lost due to parallel work and is herby reintroduced

mhsdesign commented on Neos.Workspace.Ui/Resources/Private/Fusion/Features/Review/Components/ReviewDocumentTableRow.fusion:
```diff
@@ -0,0 +1,115 @@
+      package='Neos.Neos'
+      controller='Frontend\\\Node'
+      action = 'preview'
+      // todo fixme
+      @process.addQuery = ${value + '?node=' + document.document.documentNodeAddress}
```
will be fixed via https://github.com/neos/neos-development-collection/pull/5419

mhsdesign committed:
TASK: Simplify code and remove `ContentChange` class

mhsdesign committed:
TASK: Declare all php classes in `Neos.Workspace.Ui` as internal like in the Neos.Ui

pKallert committed:
Fix: Remove ToDo

mhsdesign committed:
TASK: Refine translation of property labels

Overhauls b830ee8a2d88be1d6cc81c41667c85f1ddb5dcdd to use the translation helper to avoid duplicating code. This is done also in other places in php already.

mhsdesign committed:
BUGFIX: Allow to edit workspaces if they contain changes

Because of changes the base workspace will not be selectable
> You cannot change the base workspace of workspace with unpublished changes.

And thus also not part of the payload which crashes with a bad request (400):

> Required argument "baseWorkspace" is not set.

Instead, we make this parameter optional.

Additionally, we ensure that the base workspace is different before issuing any change.

mhsdesign committed:
TASK: Use match to validate $visibility is correct

mhsdesign committed:
TASK: Use value object for `sorting` to simplify codeflow

mhsdesign committed:
BUGFIX: Fix sorting glitch because the hot swapped WorkspaceTable on update was not sorted ASC by title

instead the order was slightly different ... neither ASC nor DESC.

to ensure we dont have to deal with that we remove the update response by just using `forward()` letting the index action handle everything correctly

mhsdesign committed:
TASK: Simplify initial workspace list rendering for the live case

mhsdesign committed:
TASK: happy new year phpstan!!!

mhsdesign reviewed (approved):
Thanks to everyone who participated here or in all the discussions! Great to have this on the finish line!!!

I tested it out and fixed some bugs:
- [BUGFIX: Allow to edit workspaces if they contain changes](https://github.com/neos/neos-development-collection/pull/5132/commits/8a94d89d74ca2eac4f137871728435706fd19b59)
- [BUGFIX: Fix sorting glitch because the hot swapped WorkspaceTable](https://github.com/neos/neos-development-collection/pull/5132/commits/bfb52da796b4d0ec499dd449258975de5389a9b3)
- [BUGFIX: Show workspaces in overview if their parent workspace is no](https://github.com/neos/neos-development-collection/pull/5132/commits/f3fe1d08a61d0bbd78eedea1b4d15f1179358bea)

And also resolved the last todos or made a followup pr. The code is now polished and can be merged!! :)

mhsdesign committed:
TASK: simplify workspace tree node fusion integration

mhsdesign committed:
BUGFIX: Show workspaces in overview if their parent workspace is not accessible

we do this by checking if a workspaces base workspace does not appear in the list of workspaces.

This is also true for the live workspace as its not part of the list so the

`entryWorkspace.baseWorkspaceName == 'live'` check is now obsolete but it shows better what we are trying to do here for the normal case :D

Also the dropdown of base workspaces must always show the current option even if it would not be re changeable in the current permission set from the current user.

pKallert reviewed (approved):
Tested locally and did not find any obvious issues
